provider/keystore/keystore.go

```diff
@@ -195,8 +238,41 @@ func (s *keystore) worker() {
 	}
 }
 
-// put stores the provided keys while assuming s.lk is already held, and
-// returns the keys that weren't present already in the keystore.
+// loadSize initializes the in-memory size counter during startup. It attempts
+// to read the persisted size from the previous session, falling back to a full
+// datastore scan if unavailable or corrupt. The size key is immediately deleted
+// after reading to ensure it remains ephemeral: only existing between clean
+// shutdown and startup. This prevents stale data if the process crashes while
+// running, and ensures the metadata key itself doesn't count toward the size.
+func (s *keystore) loadSize() {
+	sizeBytes, err := s.ds.Get(context.Background(), sizeKey)
+	if err == nil && len(sizeBytes) == 8 {
+		s.size = int(binary.BigEndian.Uint64(sizeBytes))
+		// Delete immediately to keep the key ephemeral.
+		s.ds.Delete(context.Background(), sizeKey)
+		return
+	}
+	// Size unavailable or corrupt, delete any stale metadata and perform full refresh.
+	s.ds.Delete(context.Background(), sizeKey)
+	if size, err := refreshSize(context.Background(), s.ds); err == nil {
+		s.size = size
+	} else {
+		s.logger.Error("keystore: failed to refresh size during load: ", err)
+	}
```

Comment:
https://go.dev/wiki/CodeReviewComments#indent-error-flow
```suggestion
	if err != nil || len(sizeBytes) != 8 {
		// Size unavailable or corrupt, delete any stale metadata and perform full refresh.
		s.ds.Delete(context.Background(), sizeKey)
		if size, err := refreshSize(context.Background(), s.ds); err == nil {
			s.size = size
		} else {
			s.logger.Error("keystore: failed to refresh size during load: ", err)
		}
		return
	}

	s.size = int(binary.BigEndian.Uint64(sizeBytes))
	// Delete immediately to keep the key ephemeral.
	s.ds.Delete(context.Background(), sizeKey)
```